clientoptions.go

```diff
@@ -0,0 +1,150 @@
+package influxdb
+
+import (
+	"crypto/tls"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+// Option is an option for the client config.
+type Option func(*Client) error
+
+// HTTPConfig is an https://github.com/influxdata/influxdb1-client compatible client for setting config
+// options.  This is here to make transition to go1 easy.
+// It is recommended that you set the options using the With___ functions instead.
+type HTTPConfig struct {
+	// Addr should be of the form "http://host:port"
+	// or "http://[ipv6-host%zone]:port".
+	Addr string
+
+	// Username is the influxdb username, optional.
+	Username string
+
+	// Password is the influxdb password, optional.
+	Password string
+
+	// UserAgent is the http User Agent, defaults to "InfluxDBClient" plus os and version info.
+	UserAgent string
+
+	// Timeout for influxdb writes, if set to zero, it defaults to a 20 second timeout. This is a difference from the influxdb1-client.
+	Timeout time.Duration
+
+	// InsecureSkipVerify gets passed to the http client, if true, it will
+	// skip https certificate verification. Defaults to false.
+	InsecureSkipVerify bool
+
+	// TLSConfig allows the user to set their own TLS config for the HTTP
+	// Client. If set, this option overrides InsecureSkipVerify.
+	TLSConfig *tls.Config
+
+	// Proxy configures the Proxy function on the HTTP client.
+	Proxy func(req *http.Request) (*url.URL, error)
+}
+
+// WithV1Config is an option for setting config in a way that makes it easy to convert from the old config.
+func WithV1Config(conf *HTTPConfig) Option {
+	return func(c *Client) error {
+		if conf.Addr != "" {
+			if err := WithAddress(conf.Addr)(c); err != nil {
+				return err
+			}
+		}
+		if conf.Username != "" || conf.Password != "" {
+			if err := WithUserAndPass(conf.Username, conf.Password)(c); err != nil {
+				return err
+			}
+		}
+		if conf.UserAgent != "" {
+			if err := WithUserAgent(conf.UserAgent)(c); err != nil {
+				return err
+			}
+		}
+		if conf.UserAgent != "" {
+			if err := WithUserAgent(conf.UserAgent)(c); err != nil {
+				return err
+			}
+		}
+		if conf.Timeout == 0 {
+			c.httpClient.Timeout = conf.Timeout
+		}
+		if conf.InsecureSkipVerify || conf.TLSConfig != nil || conf.Proxy != nil {
+			panic("Unimplemented")
+		}
+		return nil
+	}
+}
+
+// WithNoCompression returns an option for writing the data to influxdb without compression.
+func WithNoCompression() Option {
```

Comment:
This option conflicts with WithGZIP, what if you pass both of these?  Maybe have a single WithCompression() option?

Reply:
If you pass both, the later one overrides. I will make that clear in the comments for the `Option` type

Reply:
While a comment is good, its always best to make an interface harder to use wrong.  Maybe this option could have an argument to give it more configurability without the possibility of it being used illegally.  I realize there are still ways it can be used wrong, such as sending the option twice, but I still think it would be helpful.

Reply:
Since each type of compression, has its own options, this would complicate stuff and make it harder to use imo.  You couldn't just have a WithCompression(compressionType, level int) because different compression formats use very different levels, it isn't always an integer 0-9, and sometimes they are unrelated to anything... for example if we end up supporting ZStd, we would need to pass in a dictionary etc.  Plus, I think the chance of us adding other compression schemes is pretty high... if just because gzip isn't great on embedded devices so something like lz4 would be useful.

Gzip would require 9 different constants to use a `WithCompression(compressionType)` function for example.

How about instead I just return an error with a helpful message  if someone tries to use incompatible options?

Reply:
It's simple, just have options to your options. ;)  Just kidding of course, yeah I guess the error is best.

One thing I'm wondering now is why have WithNoCompression at all?  Isn't it the default?

Reply:
The default is to use Gzip.  I was asked to make this the default, and it makes sense to me.  The only reason to not compress is if you are resource constrained on cpu.